### PR TITLE
Mention post-fixer now correctly handles undo.

### LINKED ISSUE #8: Postfixer should remove the whole mention when it's build from two+ text nodes
Mar-27-2019 09-47-15

### LINKED ISSUE #12: Tests fail
image

### PR DESCRIPTION
**Pull request by jodator**
Fix: Mention post-fixer now correctly handles partial mentions on insert, remove and paste. Closes #12. Closes #8.

### CODE REVIEW

**Comment by scofalik** [comment]
Three things:

1. In insertions, you need to check deeply for inserted text nodes. When you inserted a paragraph, a partial mention might be in that paragraph. `Differ` returns only the top-level changes (it is optimized for downcasting). You may write it as a util (or a method on `Differ`) because it is not the first time we are having trouble with this.
2. Shift+enter and also pasting does not work correctly.
3. If we agree that mentions are always on exactly one text node (we don't allow partial styling) I am not sure if we need the "additional check for deleting last character". When is it needed? If you delete the last character in mention then, well, first of all, it would have to be a one-character-mention :D second of all, you don't need to remove the attribute from it if the whole mention is gone. So I think it is not needed?

**Comment by jodator** [comment]
>     3\. If we agree that mentions are always on exactly one text node (we don't allow partial styling) I am not sure if we need the "additional check for deleting last character". When is it needed? If you delete the last character in mention then, well, first of all, it would have to be a one-character-mention :D second of all, you don't need to remove the attribute from it if the whole mention is gone. So I think it is not needed?

AFAIR it's an edge-case when selection is after mention and you remove last character (ie `d` in `@Ted`) then the previous check wouldn't work as selection is not inside text node.

**Comment by scofalik** [comment]
Oh, ok, you meant "last" as "right-most" not "the only" :D. Yeah, I checked it and you are right.

**Comment by scofalik** [comment]
So, in this case, we also need a check for the "left-most" character and I'd like to see a more clear comment. Use "left-most" and "right-most" or "at the beginning" and "at the end" instead of "first" and "last" to make it non-ambiguous.

Also, considering the other changes, I think that maybe structuring the post-fixer this way will be better:

```js
for ( ... ) {
    if ( change.type == 'insert' ) {
        // Logic for insertions of text and non-text inside the mention.
        // And also looking for partial mentions in inserted content.
    } else if ( change.type == 'remove' ) {
        // Logic for removing of text inside, at the beginning and at the end of a mention.
    }
}
```

**Comment by jodator** [comment]
@scofalik: Ready to re-review. Two follow ups:

1. MentionUI doesn't have 100% CC
2. You can type 1 char with mention (after changing selection attributes - ie enabling typing with bold).

File: src/mentionediting.js [hunk 1]
```diff
@@ -176,3 +217,43 @@ function removePartialMentionPostFixer( writer, doc ) {
 
 	return wasChanged;
 }
+
+// Checks if node has correct mention attribute if present.
+// Returns false if node is text and has mention attribute and its text does not match expected mention text.
+//
+// @param {module:engine/model/node~Node} node a node to check
+// @returns {Boolean}
+function checkMentionAttributeOnNode( node ) {
```
[the remark this answers is not in the document]
Comment (scofalik) [reply src/mentionediting.js]: Okay I see that you use it later too in `getBrokenMentionsFromRange` so maybe `shouldFixMentionAttribute` is not the best name, so I have another proposal: `isBrokenMentionNode`.

File: src/mentionediting.js [hunk 2]
```diff
@@ -176,3 +217,43 @@ function removePartialMentionPostFixer( writer, doc ) {
...
+function* getBrokenMentionsFromRange( range ) {
+	// Check node at the left side of a range.
+	if ( !checkMentionAttributeOnNode( range.start.nodeBefore ) ) {
+		yield range.start.nodeBefore;
+	}
+
+	// Yield text nodes with broken mention from the range.
+	for ( const textProxy of range.getItems() ) {
+		if ( !checkMentionAttributeOnNode( textProxy ) ) {
+			yield textProxy.textNode;
```
Comment (scofalik) [inline src/mentionediting.js]: I understand that this is covered by unit tests but I wonder out of curiosity when this happens?

File: tests/mentionediting.js [hunk 3]
```diff
@@ -217,7 +217,45 @@ describe( 'MentionEditing', () => {
...
+				writer.setSelectionAttribute( 'bold', true );
+
+				writer.insertText( 'a', doc.selection.getAttributes(), writer.createPositionAt( paragraph, 6 ) );
+			} );
+
+			expect( getModelData( model, { withoutSelection: true } ) )
+				.to.equal( '<paragraph>foo @J<$text bold="true">a</$text>ohn bar</paragraph>' );
+		} );
+
+		it( 'should remove mention on removing a text at the beginning of a mention', () => {
```
Comment (scofalik) [inline tests/mentionediting.js]: Do we have a test for removing at the end?

File: src/mentionediting.js (hunk 2 above)
Comment (scofalik) [reply src/mentionediting.js]: I mean, if that happens for mentions that are not on the boundary of the range. Some something like `[foo<mention />bar]`.

Comment (jodator) [reply src/mentionediting.js]: So this is used in attribute post-fixer:

```html
fo[o <$text mention="{...}">men]tion</$text>
```

So when the range on which bold is applied is ends inside a mention. Then I want to get that dangling mention node on the right(left) side of range.end(start).

File: src/mentionediting.js [hunk 4]
```diff
@@ -140,34 +141,74 @@ function selectionMentionAttributePostFixer( writer, doc ) {
...
-			const nodeBefore = change.position.nodeBefore;
+		// Inserted inline elements might break mention.
+		if ( change.type == 'insert' && schema.isInline( change.name ) ) {
+			checkAndFix( position.nodeBefore );
 
-			if ( nodeBefore && nodeBefore.hasAttribute( 'mention' ) ) {
-				const text = nodeBefore.data;
-				const mention = nodeBefore.getAttribute( 'mention' );
+			const nodeAfterInserted = position.nodeAfter && position.nodeAfter.nextSibling;
+			checkAndFix( nodeAfterInserted );
```
[the remark this answers is not in the document]
Comment (jodator) [reply src/mentionediting.js]: don't be sorry it's nicer now :)

File: src/mentionediting.js [hunk 5]
```diff
@@ -140,34 +141,74 @@ function selectionMentionAttributePostFixer( writer, doc ) {
 // @param {module:engine/model/writer~Writer} writer
 // @param {module:engine/model/document~Document} doc
 // @returns {Boolean} Returns true if selection was fixed.
-function removePartialMentionPostFixer( writer, doc ) {
+function removePartialMentionPostFixer( writer, doc, schema ) {
 	const changes = doc.differ.getChanges();
 
 	let wasChanged = false;
 
+	if ( !changes.length ) {
```
[the remark this answers is not in the document]
Comment (jodator) [reply src/mentionediting.js]: nah - probably some debug left-over.

File: src/mentionediting.js [hunk 6]
```diff
@@ -140,34 +141,74 @@ function selectionMentionAttributePostFixer( writer, doc ) {
...
+
+			// Occurs on pasting inside text node with mention;
+			const nodeAfterInsertedTextNode = position.textNode && position.textNode.nextSibling;
+			checkAndFix( nodeAfterInsertedTextNode );
+			checkAndFix( position.nodeBefore );
+			checkAndFix( position.nodeAfter );
+		}
+
+		// Check text nodes on inserted elements (might occur when splitting paragraph on enter key).
+		if ( change.name != '$text' && change.type == 'insert' && schema.checkChild( change.name, '$text' ) ) {
```
[the remark this answers is not in the document]
Comment (jodator) [reply src/mentionediting.js]: But this is fixed by upcast converter. This check is for splitting parent element with mention when split occurred inside text with mention. As such there is no need to check other cases then the elements that allow text inside.

File: src/mentionediting.js [hunk 7]
```diff
@@ -140,34 +141,74 @@ function selectionMentionAttributePostFixer( writer, doc ) {
...
+			for ( const child of insertedNode.getChildren() ) {
+				checkAndFix( child );
 			}
 		}
 
-		// Additional check for deleting last character of a text node.
-		if ( change.type == 'remove' ) {
-			const nodeBefore = change.position.nodeBefore;
+		// Inserted inline elements might break mention.
+		if ( change.type == 'insert' && schema.isInline( change.name ) ) {
```
[the remark this answers is not in the document]
Comment (jodator) [reply src/mentionediting.js]: no it explicitly checks for inserted in-line elements like `<softBreak>` which may break text (with mention.

File: tests/mentionediting.js (hunk 3 above)
Comment (jodator) [reply tests/mentionediting.js]: it is covered by `removing text in the middle a mention` but I might add one explicitly.

Comment (jodator) [reply tests/mentionediting.js]: Ugh.. its right there, few lines below :P

Comment (jodator) [reply tests/mentionediting.js]: Selection_279

Comment (scofalik) [reply tests/mentionediting.js]: Fine, I checked only the changed code and didin't see it.

**Comment by jodator** [comment]
@scofalik I've resolved all the issues from your comments.

I would leave those optimizations are they checks cases that might break mention during editing.

The pasting part of nested nodes is done by upcast converter - it does not allow partial mentions.

I've added tests for undo integration with attribute post-fixer.

File: src/mentionediting.js (hunk 2 above)
Comment (scofalik) [reply src/mentionediting.js]: So isn't it that we are always interested in nodes next to range boundaries?

Comment (jodator) [reply src/mentionediting.js]: When applying attribute on this range yes.

**Comment by scofalik** [comment]
> The pasting part of nested nodes is done by upcast converter - it does not allow partial mentions.

Typically, I prefer having general solutions that cover the whole problem instead of directly solving the predicted cases. This is safer (and, in my opinion, correct) way to go. For example, I just had to fix a bug in view writer because some behavior was predicted only for one case while it could simply be inside a `for` loop following the special cases and applying wrapping/unwrapping to all children recursively.

This is why I have mixed feelings about it. I don't want to introduce redundant code that might never be used. But I also don't want to introduce potentially bad code. For now, if I didn't miss anything, the code is fine. However, if there is ever a feature (widget) that breaks blockquote, there will be a problem:

```html
<blockQuote><paragraph><$text mention="jodator">@jod[]ator</$text></paragraph></blockQuote>
```

Paste the widget that can be only in the root at `[]`:

```html
<blockQuote><paragraph><$text mention="jodator">@jod</$text></paragraph></blockQuote>
<hypotethicalWidget />
<blockQuote><paragraph><$text mention="jodator">ator</$text></paragraph></blockQuote>
```

Or when a 3rd party plugin will simply split both `paragraph` and `blockQuote`. Or if we will support more nested structure and for some reason, we will break multiple blocks at once.

Anyway, it is hypothetical for now. I tried to think about cases in OT and block lists (the thing we plan to introduce, where you will be able to, for example, add multiple paragraphs, headers, maybe tables to list items) but I don't know if anything there will break this behavior.

So as I said, I have mixed feelings. I summon @Reinmar .

**Comment by jodator** [comment]
> Paste the widget that can be only in the root at `[]`:

OK valid TC. I've manged to do it with 
```
blockquote > paragraph > mention
```
and split:
```
editor.model.change( writer => {
 writer.split( editor.model.document.selection.focus, editor.model.document.getRoot() );
} );
```
I'll update the PR for deep check.

**Pull request merged**